conda_auth/handlers/base.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Callable
+
+import conda.base.context
+import keyring
+import requests
+from conda.gateways.connection.session import get_session
+from conda.models.channel import Channel
+
+from ..exceptions import CondaAuthError
+
+INVALID_CREDENTIALS_ERROR_MESSAGE = "Provided credentials are not correct."
+
+
+class AuthManager(ABC):
+    """
+    Defines an interface for auth handlers to use within plugin
+    """
+
+    def __init__(self, context: conda.base.context.Context, cache: dict | None = None):
+        """
+        Optionally set a cache to use and configuration parameters to retrieve from
+        ``conda.base.context.context.channel_settings``.
+        """
+        self._context = context
+        self._cache = {} if cache is None else cache
+
+    def get_action_func(self) -> Callable[[str], None]:
+        """Return a callable to be used as the action function for the pre-command plugin hook"""
+
+        def action(command: str):
+            for settings in self._context.channel_settings:
+                if channel := settings.get("channel"):
+                    channel = Channel(channel)
+                    # Only attempt to authenticate for actively used channels
+                    if (
+                        channel.canonical_name in self._context.channels
+                        and settings.get("auth") == self.get_auth_type()
+                    ):
+                        self.authenticate(channel, settings)
+
+        return action
```

Comment:
Could this be boiled down to not return a callable, but BE the callable?

```suggestion
    def action_func(self, command: str):
        """The callable to be used as the action function for the pre-command plugin hook"""

        for settings in self._context.channel_settings:
            if channel := settings.get("channel"):
                channel = Channel(channel)
                # Only attempt to authenticate for actively used channels
                if (
                    channel.canonical_name in self._context.channels
                    and settings.get("auth") == self.get_auth_type()
                ):
                    self.authenticate(channel, settings)

```

Reply:
Yeah, that makes sense.

I think I did this before just making this object itself a singleton. Now that it's a singleton, we could just directly invoke that method.